fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,784 @@
+use std::collections::BTreeSet;
+use std::fs::{self, File};
+use std::io::Write;
+use std::time::SystemTime;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use crate::base_storage::{BaseStorage, SyncStatus};
+use crate::monoid::Monoid;
+use data_error::{ArklibError, Result};
+
+/// Represents a folder storage system that persists data to disk.
+pub struct FolderStorage<K, V> {
+    /// Label for logging
+    label: String,
+    /// Path to the underlying folder where data is persisted
+    path: PathBuf,
+    /// `modified` can be used to track the last time a file was modified in memory.
+    /// where the key is the path of the file inside the directory.
+    modified: BTreeMap<K, SystemTime>,
+    /// `written_to_disk` can be used to track the last time a file written or read from disk.
+    /// where the key is the path of the file inside the directory.
+    written_to_disk: BTreeMap<K, SystemTime>,
+    data: BTreeMap<K, V>,
+    /// Temporary store for deleted keys until storage is synced
+    deleted_keys: BTreeSet<K>,
+}
```

Comment:
To simplify, we could combine `modified` and `written_to_disk` into a single `BTreeMap<K, (SystemTime, SystemTime)>`, which we can call `timestamps`. This would avoid duplicating keys in memory.

We could even take it further and merge this with the `data` map into one struct, with a field like `deleted` to indicate whether an entry is deleted or not.

Since we don’t serialize directly but loop over the key-value pairs, I don’t think this should be a problem.

Reply:
Yeah, storing modified and written_to_disk timestamps in the same mapping seems like a better approach since it brings all the metadata together. However, I think it's better to keep the actual value in a separate map as it keeps metadata and actual data separate from each other.

I believe this will be helpful when creating bindings for JNI. If they were stored in a single struct and we don't want to expose internal metadata through Java interface (for encapsulation, perf reasons) then we'd have to "project" the necessary fields out of the struct making the solution more complex.

I think the timestamps and data mapping is a good middle ground.